File: common/src/main/java/net/impleri/playerskills/utils/RegistrationType.java
package net.impleri.playerskills.utils;

import net.minecraft.core.Registry;
import net.minecraft.resources.ResourceKey;
import net.minecraft.resources.ResourceLocation;
import net.minecraft.tags.TagKey;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.util.function.Consumer;

public class RegistrationType<T> {
    @Nullable
    private final String namespace;
    @Nullable
    private final ResourceLocation name;
    @Nullable
    private final TagKey<T> tag;

    public RegistrationType(String value, ResourceKey<Registry<T>> registryKey) {
        if (value.trim().startsWith("@")) {
            namespace = value.substring(1);
            name = null;
            tag = null;
        } else if (value.trim().startsWith("#")) {
            var tagKey = value.substring(1);

            tag = TagKey.create(registryKey, SkillResourceLocation.ofMinecraft(tagKey));
            namespace = null;
            name = null;
        } else if (value.trim().endsWith(":*")) {
            namespace = value.substring(0, value.indexOf(":"));
            name = null;
            tag = null;
        } else {
            name = SkillResourceLocation.ofMinecraft(value);
            namespace = null;
            tag = null;
        }
    }

    public void ifTag(@NotNull Consumer<TagKey<T>> consumer) {
        if (tag != null) {
            consumer.accept(tag);
        }
    }

    public void ifName(@NotNull Consumer<ResourceLocation> consumer) {
        if (name != null) {
            consumer.accept(name);
        }
    }

    public void ifNamespace(@NotNull Consumer<String> consumer) {
        if (namespace != null) {
            consumer.accept(namespace);
        }
    }
}

File: common/src/main/kotlin/net/impleri/playerskills/utils/RegistrationType.kt
package net.impleri.playerskills.utils

import net.impleri.playerskills.utils.SkillResourceLocation.ofMinecraft
import net.minecraft.core.Registry
import net.minecraft.resources.ResourceKey
import net.minecraft.resources.ResourceLocation
import net.minecraft.tags.TagKey

class RegistrationType<T>(value: String, registryKey: ResourceKey<Registry<T>>) {
  private var namespace: String? = null
  private var name: ResourceLocation? = null
  private var tag: TagKey<T>? = null

  init {
    if (value.trim { it <= ' ' }.startsWith("@")) {
      namespace = value.substring(1)
      name = null
      tag = null
    } else if (value.trim { it <= ' ' }.startsWith("#")) {
      val tagKey = value.substring(1)
      tag = TagKey.create(registryKey, ofMinecraft(tagKey))
      namespace = null
      name = null
    } else if (value.trim { it <= ' ' }.endsWith(":*")) {
      namespace = value.substring(0, value.indexOf(":"))
      name = null
      tag = null
    } else {
      name = ofMinecraft(value)
      namespace = null
      tag = null
    }
  }

  fun ifTag(consumer: (TagKey<T>) -> Unit) {
    tag?.let { consumer(it) }
  }

  fun ifName(consumer: (ResourceLocation) -> Unit) {
    name?.let { consumer(it) }
  }

  fun ifNamespace(consumer: (String) -> Unit) {
    namespace?.let { consumer(it) }
  }
}
